src/coreclr/tests/src/JIT/Directed/IL/PInvokeTail/PInvokeTail.il

```diff
@@ -28,78 +27,66 @@
 
 } // end of class 'Wrapper'
 
-.method public static int32 ?callputs1@@YAHXZ() il managed
-{
-  .maxstack  1
-
-  IL_0000:  ldstr bytearray(48 65 6C 6C 6F 20 57 6F 72 6C 64 21 00 00 00 00) //ldptr      D_00003000
-  IL_0005:  tail. call       int32 puts(int8*)
-  IL_000a:  ret
-} // end of global method '?callputs1@@YAHXZ'
+.method public static pinvokeimpl("msvcrt.dll" cdecl) int32 puts(int8* A_0) il managed preservesig{ }
 
-.method public static int32 ?callputs2@@YAHXZ() il managed
+.method public static int32 callputs1() il managed
 {
-  .maxstack  1
-
-  IL_0000:  ldstr bytearray(48 65 6C 6C 6F 20 57 6F 72 6C 64 21 00 00 00 00) //"  " //ldptr      D_00003010
-  IL_0005:  tail. call       int32 Wrapper::puts(int8*)
-  IL_000a:  ret
-} // end of global method '?callputs2@@YAHXZ'
+    .maxstack 1
+    .locals init (char&)
+
+    ldstr bytearray(48 65 6C 6C 6F 20 57 6F 72 6C 64 21 00 00 00 00) // "Hello World!"
+    call instance char& modreq([mscorlib]System.Runtime.InteropServices.InAttribute) [mscorlib]System.String::GetPinnableReference()
+    stloc.0
+    ldloc.0
+    conv.i
+    tail. call int32 puts(int8*)
```

Comment:
This is still passing unpinned pointer into a PInvoke. It will have unpredictable behavior e.g. with GC stress if the string gets moved while the PInvoke is executing.

Reply:
Oops. Forgot the `pinned`. I was doing a bit of experimentation here trying to learn some IL.